FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+- range-based vs. point-wise selection
+- exact vs. inexact lookups
+
+Whether or not a `XarrayIndex` subclass supports each of the features listed above should be either declared explicitly via a common API or left to the implementation. An `XarrayIndex` subclass may encapsulate more than one underlying object used to perform the actual indexing. Such "meta" index would typically support a range of features among those mentioned above and would automatically select the optimal index object for a given indexing operation.
+
+Every `XarrayIndex` subclass must at least implement two methods:
+
+- One `build` method that takes one or more Dataset/DataArray coordinates and that returns the object(s) that will be used for the actual indexing (e.g., `pandas.Index`, `scipy.spatial.KDTree`, etc.)
+- One `query` method that takes label-based indexers as argument and that returns the corresponding position-based indexers.
+
+These two methods may accept additional keyword arguments passed to the underlying index object constructor or query methods.
+
+There are potentially other properties / methods that an `XarrayIndex` subclass must/should/may implement, e.g.,
+
+- An `indexes` property to access the underlying index object(s) wrapped by the `XarrayIndex` subclass
+- A `data` property to access index's data and map it to coordinate data (see [Section 4](#4-indexvariable))
+- a `__getitem__()` implementation to propagate the index through DataArray/Dataset indexing operations
+- `equals()`, `union()` and `intersection()` methods for data alignment (see [Section 2.6](#26-using-indexes-for-data-alignment))
+- Xarray coordinate getters (see [Section 2.2.4](#224-implicit-coodinates))
+- A method that may return a new index and that will be called when one of the corresponding coordinates is dropped from the Dataset/DataArray (multi-coordinate indexes)
+- `encode()`/`decode()` methods that would allow storage-agnostic serialization and fast-path reconstruction of the underlying index object(s) (see [Section 2.8](#28-index-encoding))
+- One or more "non-standard" methods or properties that could be leveraged in Xarray 3rd-party extensions like Dataset/DataArray accessors (see [Section 2.7](#27-using-indexes-for-other-purposes))
+
+The `XarrayIndex` API has still to be defined in detail.
+
+Xarray should provide a minimal set of built-in index wrappers (this could be reduced to the indexes currently supported in Xarray, i.e., `pandas.Index` and `pandas.MultiIndex`). Other index wrappers may be implemented in 3rd-party libraries (recommended). The `XarrayIndex` base class should be part of Xarray's public API.
+
+#### 2.1.1 Index discoverability
+
+For better discoverability of Xarray-compatible indexes, Xarray could provide some mechanism to register new index wrappers, e.g., something lire [xoak's `IndexRegistry`](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.IndexRegistry.html#xoak.IndexRegistry).
```

Comment:
The [numcodecs registry](https://numcodecs.readthedocs.io/en/stable/registry.html) could be another possible example.